### ISSUE TITLE
APISec: Extend with extra documentation and introduce metadata information to all API endpoint

### ISSUE THREAD

**Issue by emmdim** [vocdoni/interoperability#70]
Following from  https://github.com/vocdoni/interoperability/issues/65

1)  @selankon  please, extend it to other API endpoints and types.
As mentioned in https://github.com/vocdoni/vocdoni-node/pull/944, 

2) @altergui please could you automate the generation of swagger using a separate github workflow triggered by PUSH event? Once completed, I @chilcano can add github jobs to execute the OWASP API security review and publish the findings in there.

**Comment by selankon** [comment]
I have a big stopper when documenting protobuff generated classes. For example with this method:

```
/chain/transactions/{height}/{index}
```

This endpoint return a object `tx` that can get the shape of any of the existing transactions. This transactions are defined on protobuff but I am not sure of how are they generated, how to import it etc...

A solution could be to generate this docs manually and then add it using a shell script after generations. 

This is a mid blocker, expecting to find a solution.

**Comment by selankon** [comment]
TLDR

If I use the protobuff model of `models.Tx`, the generated swagger is:

```yaml
  models.Tx:
    properties:
      payload:
        description: "<parsed description>"
    type: object
```

So it cannot take the protobuf models properly (it seem has problems with the `oneof` protobuf annotation). I am still trying to find a solution, some options I am considering:

- Write transactions objects by hand and merge it into the yaml using [`bash + yq`](https://github.com/kislyuk/yq) or using a [python script](https://stackoverflow.com/a/63582180) rather than a `sed` or `awk` script due to it scalability.
- Try to generate the protobuff models of transaction (`Tx_Vote`, `Tx_Admin`...) using some swaggo hack. For example writing a dummy api.go with this models, generate them using swaggo and merging the yml. 
- ... yet thinking about ...


Some discarded options

- Change generator to https://github.com/go-swagger/go-swagger . I discarded this option because I didn't saw any evidence that this generator has better implementation for protobuf `oneof` models.

**Comment by selankon** [comment]
Update

The `oneof` feature is what we need to describe the transaction by block height and index response.  It can return a transaction of type `Tx_Vote`, `Tx_Admin`... Whatever.

After more research, i found that the `oneof` feature is only for Openapi 3.0, which [is not yet supported by swaggo ](https://github.com/swaggo/swag/issues/1363#issuecomment-1514999343)  and is [not supported by the swagger 2.0](https://github.com/swaggo/swag/issues/1363#issuecomment-1296472499) neither.

The most similar behavior of the `oneof` feature on swagger 2.0, seems to be the [discriminator field](https://github.com/swaggo/swag/issues/1278#issuecomment-1232574890), described [here](https://swagger.io/specification/v2/#composition-and-inheritance-polymorphism). 

BTW, is not exactly what we need, and seems that swaggo [is on the way to give support](https://github.com/swaggo/swag/issues/386) to openapi 3.0. 

Meanwhile, I found that this library, https://github.com/getkin/kin-openapi, supports Openapi v3, but I can find how to be used as easy as swaggo. 

To generate automatically the swagger models for the transactions, I created this minimum example that generates a yml with the models

```go
package protobufmodels

import (
	_ "go.vocdoni.io/proto/build/go/models"
	_ "google.golang.org/protobuf/proto"
)

// generate transaction models
//
//	@Success		200	{object}	models.Tx_Vote
//	@Success		200	{object}	models.Tx_NewProcess
func chainTxHandler() {
}
```

This can be used to generate the models and then merge them to the swagger, modifying the transaction response using the discriminator field or similar.

**Comment by selankon** [comment]
After deep research, I concluded with this:  https://github.com/vocdoni/vocdoni-node/blob/4b7d32aead7f0410bdd83462458b2d02acd4d4c4/api/docs/README.md

Could you @altergui and @chilcano  take a look on it?

**Comment by selankon** [comment]
This is closed by https://github.com/vocdoni/vocdoni-node/pull/944  and https://github.com/vocdoni/vocdoni-node/pull/1036

- It generates swaggo docs on every merge to main
- All endpoints are documented and tested

Imo  this is closed

**Comment by altergui** [comment]
:partying_face: :rocket:
